contracts/UnionZap.sol

```diff
@@ -319,64 +361,211 @@ contract UnionZap is Ownable, UnionBase {
+                // to update the cvxCRV distributor's stake function
+                if (_outputToken == CRV_TOKEN) {
+                    // convert all CRV to cvxCRV
+                    _toCvxCrv(minAmounts[i], lock);
+                } else {
+                    // slippage check
+                    assert(
+                        IERC20(_outputToken).balanceOf(address(this)) >
+                            minAmounts[i]
+                    );
+                }
+            }
         }
+    }
 
-        if (_gasCostInCrv > 0) {
-            uint256 _gasEth = _swapCrvToEth(
-                _swapCvxCrvToCrv(_gasCostInCrv, address(this))
-            );
-            (bool success, ) = (msg.sender).call{value: _gasEth}("");
-            require(success, "ETH transfer failed");
+    /// @notice Deposits rewards to their respective merkle distributors
+    /// @param weights - weights of output assets (cvxCRV, FXS, CVX...)
+    function distribute(uint16[] calldata weights)
+        public
+        onlyOwner
+        validWeights(weights)
+    {
+        for (uint256 i; i < weights.length; ++i) {
+            if (weights[i] > 0) {
+                address _outputToken = outputTokens[i];
+                address _distributor = tokenInfo[_outputToken].distributor;
+                IMerkleDistributorV2(_distributor).freeze();
+                // edge case for CRV as we gotta keep using existing distributor
+                if (_outputToken == CRV_TOKEN) {
+                    _outputToken = CVXCRV_TOKEN;
+                }
+                uint256 _balance = IERC20(_outputToken).balanceOf(
+                    address(this)
+                );
+                // transfer to distributor
+                IERC20(_outputToken).safeTransfer(_distributor, _balance);
+                // stake
+                IMerkleDistributorV2(_distributor).stake();
+                emit Distributed(_balance, _outputToken, _distributor);
+            }
         }
-        emit Distributed(_netDeposit, _gasCostInCrv, _locked);
+    }
+
+    /// @notice Swaps all bribes, adjust according to output token weights and distribute
+    /// @dev Wrapper over the swap, adjust & distribute function
+    /// @param claimParams - an array containing the info necessary to claim
+    /// @param routerChoices - the router to use for the swap
+    /// @param claimBeforeSwap - whether to claim on Votium or not
+    /// @param gasRefund - tx gas cost to refund to caller (ETH amount)
+    /// @param weights - weight of output assets (cvxCRV, FXS, CVX...) in bips
+    /// @param minAmounts - min amount out of each output token (cvxCRV for CRV)
+    function processIncentives(
+        IMultiMerkleStash.claimParam[] calldata claimParams,
+        uint256 routerChoices,
+        bool claimBeforeSwap,
+        bool lock,
+        uint256 gasRefund,
+        uint16[] calldata weights,
+        uint256[] calldata minAmounts
+    ) external onlyOwner validWeights(weights) {
```

Comment:
Referring to this comment: https://github.com/convex-community/union_contracts/pull/12/files#r854685059.

This method would see major gas savings since there are a total of 4 `validateWeights` calls (`processIncentives`, `swap`, `adjust`, and `distribute`)

Reply:
Partially fixed in 4162348960d6910c7af1ca572651d043684cba14
No real way to avoid all the redundancies since the `swap`, `adjust` and `distribute` functions should be callable on their own (and will require a validity check for the wait)